base-station/ir_interface.py

```diff
@@ -47,15 +45,377 @@
 
 """
 
+"""
+My implementation
+I expect badge will send ir packet to base station actively.
+
+USB Serial Commands:
+
+Nesseary Fields:
+Preamble - 8 bytes. 0xD5 0x55 0x55 0x55 0x55 0x55 0x55 0x55
+
+PacketType - 1 byte
+    * 0x01 for QueueTxBufferRequest
+    * 0x81 for QueueTxBufferResponse
+    * 0x02 for RetrieveRxBufferRequest 
+    * 0x82 for RetrieveRxBufferResponse
+    * 0x03 for GetStatusRequest
+    * 0x83 for GetStatusResponse
+    * 0x04 for SendStatusRequest
+    * 0x84 for SendStatusResponse
+    * 0x05 for PopRxBufferRequest (for duplex mode)
+    * 0x85 for PopRxBufferResponse
+
+Optional Fields:
+PacketSequence - 1 byte. A random byte to denote which response maps to which request.
+    The response should have the same packet sequence as the request.
+
+PacketSize - 1 byte.
+
+PacketData - x bytes
+
+Api of each packet type:
+QueueTxBufferRequest: (Computer to Badge)
+1 byte - PacketSequence.
+1 byte - PacketSize.
+x bytes - The entire buffer data to queue for transmit.
+
+QueueTxBufferResponse: (Badge to Computer)
+1 byte - PacketSequence.
+1 byte - 0x01 for buffer full, 0x02 for success.
+
+RetrieveRxBufferRequest: (Computer to Badge)
+None
+
+RetrieveRxBufferResponse: (Badge to Computer)
+1 byte - PacketSize. 0x00 for buffer empty.
+x bytes for the entire buffer received.
+
+GetStatusRequest: (Computer to Badge)
+None
+
+GetStatusResponse: (Badge to Computer)
+1 byte - 0x01 for failure, 0x02 for success.
+1 byte - Status:
+            * 0x01: Set for not ready to receive, all received buffers has been populated with received packet.
+            * 0x02: Set for ready to receive, there are empty received buffers exist on badge.
+
+SendStatusRequest: (Badge to Computer)
+1 byte - The status, same as above.
+
+SendStatusResponse: (Computer to Badge)
+None
+
+PopRxBufferRequest: (Badge to Computer)
+1 byte - PacketSequence.
+1 byte - PacketSize.
+x bytes - The entire buffer data to queue for transmit.
+
+PopRxBufferResponse: (Computer to Badge)
+1 byte - PacketSequence.
+1 byte - 0x01 for failure, 0x02 for success.
+
+"""
+
+import serial
+import enum
+from config import Config
+import asyncio
+import time
+from dataclasses import dataclass
+from typing import Optional
+
+# PacketType enum
+class PT(enum.Enum):
+    QTQ = b'\x01' # QueueTxBufferRequest
+    QTR = b'\x81' # QueueTxBufferResponse
+    RRQ = b'\x02' # RetrieveRxBufferRequest
+    RRR = b'\x82' # RetrieveRxBufferResponse
+    GSQ = b'\x03' # GetStatusRequest
+    GSR = b'\x83' # GetStatusResponse
+    SSQ = b'\x04' # SendStatusRequest
+    SSR = b'\x84' # SendStatusResponse
+    PRQ = b'\x05' # PopRxBufferRequest
+    PRR = b'\x85' # PopRxBufferResponse
+
+# Packet content format
+@dataclass
+class Packet:
+    seq: Optional[bytes] = None
+    packet_size: Optional[bytes] = None
```

Comment:
Optional: Consider making this packet_size_raw or make it int to avoid comparing/using it as bytes/string when int is expected.